src/client/datascience/jupyter/kernels/helpers.ts

```diff
@@ -73,10 +76,18 @@ export function getKernelId(spec: IJupyterKernelSpec, interpreter?: PythonEnviro
     // & sometimes the kernelspec might not have the version number (if it wasn't available at the time of generation of spec)
     // See getInterpreterKernelSpecName for details of this logic.
     let specName = spec.name;
+    const kernelRegistrationKind = getKernelRegistrationInfo(spec);
     // eslint-disable-next-line @typescript-eslint/no-use-before-define
-    if (isKernelRegisteredByUs(spec) && specName.includes(autoGeneratedKernelNameIdentifier)) {
+    if (kernelRegistrationKind && specName.includes(autoGeneratedKernelNameIdentifier)) {
         // eslint-disable-next-line @typescript-eslint/no-use-before-define
         specName = specName.substring(specName.indexOf(autoGeneratedKernelNameIdentifier));
+        // When users create kernelspecs, we need to include the name of custom kernelspecs in the id.
+        if (kernelRegistrationKind === 'registeredByNewVersionOfExtForCustomKernelSpec') {
+            const originalSpecFile = spec.metadata?.jupyter?.originalSpecFile || spec.metadata?.originalSpecFile;
+            if (originalSpecFile) {
+                specName = `${specName}#${path.basename(path.dirname(originalSpecFile))}`;
```

Comment:
Adds support for custom kernelspecs that users create (we'll give them a unique id, Id will contain the original name of the kernel the user created).